ddtrace/sampler.py

```diff
@@ -92,3 +103,266 @@ def set_sample_rate_by_service(self, rate_by_service):
+
+        if tags is not self.NO_RULE and not isinstance(tags, dict):
+            raise TypeError('SamplingRule(tags={!r}) must be a dict')
+        self.tags = tags
+
+    @property
+    def sample_rate(self):
+        return self._sample_rate
+
+    @sample_rate.setter
+    def sample_rate(self, sample_rate):
+        self._sample_rate = sample_rate
+        self._sampling_id_threshold = sample_rate * MAX_TRACE_ID
+
+    def _pattern_matches(self, prop, pattern):
+        # If the rule is not set, then assume it matches
+        # DEV: Having no rule and being `None` are different things
+        #   e.g. ignoring `span.service` vs `span.service == None`
+        if pattern is self.NO_RULE:
+            return True
+
+        # If the pattern is callable (e.g. a function) then call it passing the prop
+        #   The expected return value is a boolean so cast the response in case it isn't
+        if callable(pattern):
+            try:
+                return bool(pattern(prop))
+            except Exception as e:
+                log.warning('%r pattern %r failed with %r: %s', self, pattern, prop, e)
+                # Their function failed to validate, assume it is a False
+                return False
+
+        # The pattern is a regular expression and the prop is a string
+        if isinstance(pattern, pattern_type):
+            try:
+                return bool(pattern.match(str(prop)))
+            except (ValueError, TypeError) as e:
+                # This is to guard us against the casting to a string (shouldn't happen, but still)
+                log.warning('%r pattern %r failed with %r: %s', self, pattern, prop, e)
+                return False
+
+        # Exact match on the values
+        return prop == pattern
+
+    def _tags_match(self, span):
+        # No rule was set, then assume it matches
+        if self.tags is self.NO_RULE:
+            return True
+
+        # No tags were provided to check, so it passes
+        if not self.tags:
+            return True
+
+        # Check patterns from all tags
+        return all(
+            self._pattern_matches(span.get_tag(tag), pattern)
+            for tag, pattern in self.tags.items()
+        )
+
+    def matches(self, span):
+        """
+        Return if this span matches this rule
+
+        :param span: The span to match against
+        :type span: :class:`ddtrace.span.Span`
+        :returns: Whether this span matches or not
+        :rtype: :obj:`bool`
+        """
+        # make sure service, name, and resource match
+        if not all(
+            self._pattern_matches(prop, pattern)
+            for prop, pattern in [
+                    (span.service, self.service),
+                    (span.name, self.name),
+                    (span.resource, self.resource),
+            ]
+        ):
+            return False
+
+        # Make sure tags match
+        if not self._tags_match(span):
+            return False
+
+        return True
```

Comment:
So:
```python
return self._tags_match(span)
```

In the end:

```python
return all(
            self._pattern_matches(prop, pattern)
            for prop, pattern in [
                    (span.service, self.service),
                    (span.name, self.name),
                    (span.resource, self.resource),
            ]
        ) and self._tags_match(span)
```

looks clearer.

Reply:
thanks, good catch